datafusion/physical-expr/src/window/window_frame_state.rs

```diff
@@ -0,0 +1,972 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module provides utilities for window frame index calculations depending on the window frame mode:
+//! RANGE, ROWS, GROUPS.
+
+use arrow::array::ArrayRef;
+use arrow::compute::kernels::sort::SortOptions;
+use datafusion_common::bisect::{bisect, find_bisect_point};
+use datafusion_common::{DataFusionError, Result, ScalarValue};
+use datafusion_expr::{WindowFrame, WindowFrameBound, WindowFrameUnits};
+use std::cmp::min;
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+/// This object stores the window frame state for use in incremental calculations.
+#[derive(Debug)]
+pub enum WindowFrameContext<'a> {
+    // ROWS-frames are inherently stateless:
+    Rows(&'a Arc<WindowFrame>),
```

Comment:
As a matter of style, I find `&Arc<WindowFrame>` strange. I would expect either an owned Arc (aka increment the ref count):

```suggestion
    Rows(Arc<WindowFrame>),
```

Or else just a reference:

```suggestion
    Rows(&'a WindowFrame),
```


I don't understand the need to have a reference to the Arc